lib/elm-test.js

```diff
@@ -305,77 +304,53 @@ function infoLog(msg) {
   }
 }
 
+// It's important to globify all the arguments.
+// On Bash 4.x (or zsh), if you give it a glob as its last argument, Bash
+// translates that into a list of file paths. On bash 3.x it's just a string.
+// Ergo, globify all the arguments we receive.
+const isMake = args._[0] === 'make';
+const testFileGlobs = isMake ? args._.slice(1) : args._;
+const testFilePaths = resolveGlobs(testFileGlobs);
 const projectRootDir = process.cwd();
+const generatedCodeDir = Compile.getGeneratedCodeDir(projectRootDir);
+const hasBeenGivenCustomGlobs = testFileGlobs.length > 0;
 
-if (args._[0] === 'make') {
-  const files = args._.slice(1);
+const elmJsonPath = path.resolve(path.join(projectRootDir, 'elm.json'));
+let projectElmJson;
 
-  // It's important to globify all the arguments.
-  // On Bash 4.x (or zsh), if you give it a glob as its last argument, Bash
-  // translates that into a list of file paths. On bash 3.x it's just a string.
-  // Ergo, globify all the arguments we receive.
-  const fileGlobs = files.length > 0 ? files : [];
-  const testFilePaths = resolveGlobs(files);
-  const hasBeenGivenCustomGlobs = fileGlobs.length > 0;
-  const elmJsonPath = path.resolve(path.join(projectRootDir, 'elm.json'));
+try {
+  projectElmJson = fs.readJsonSync(elmJsonPath);
+} catch (err) {
+  console.error('Error reading elm.json: ' + err.message);
+  throw process.exit(1);
```

Comment:
Do we have to crash if we are in watch mode and the project elm json is invalid?

Reply:
And what happens if someone edits their elm.json while we are watching tests>

Reply:
I didn’t think about that case! This could for sure be a future enhancement.

Currently this is done outside of the watch loop so it would require more changes than I’m comfortable with at the moment. This _looks_ like added code but it’s in fact just moved so were keeping the same behavior as before regarding elm.json at least.

Reply:
Before this PR `generateElmJson` read the elm.json file from disk everything. Would that mean that it would detect changes in your elm.json during watch mode?

Reply:
Ah, good catch, didn’t think about that change. I’ll have to investigate.

Reply:
Turns out `Generate.generateElmJson` is outside the watch loop too (and was before this PR as well). And moving it into the loop brings more issues:

- We don’t watch `elm.json` so you need to edit some .elm file to trigger.
- If you edit `"source-directories"` we need to recalculate which globs the watcher should look at, which is also outside the loop.
- Reading elm.json and running `elm-json` currently does `process.exit(1)` on failure which would need to be refactored.
- We basically need to restart the whole thing if elm.json changes.

My conclusion is:

- This PR has not changed behavior.
- Adding support for elm.json changes could be a nice future improvement.

Reply:
> My conclusion is:
> 
> * This PR has not changed behavior.

Good enough for me!